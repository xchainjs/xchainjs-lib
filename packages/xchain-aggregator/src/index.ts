export { Aggregator } from './aggregator'
export {
  IProtocol,
  QuoteSwapParams,
  QuoteSwap,
  TxSubmitted,
  SwapHistory,
  SwapHistoryParams,
  SwapResume,
  Config,
} from './types'
