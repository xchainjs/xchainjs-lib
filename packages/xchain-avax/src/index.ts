// Export all elements from the 'client' module
import { ClientKeystore } from './ClientKeystore'

export { ClientKeystore, ClientKeystore as Client } from './ClientKeystore'
export { ClientLedger } from './ClientLedger'

// Export all elements from the 'const' module
export * from './const'

export default ClientKeystore
