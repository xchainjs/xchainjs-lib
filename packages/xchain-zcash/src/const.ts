import { ExplorerProvider, Network } from '@xchainjs/xchain-client'
import { Asset, AssetType } from '@xchainjs/xchain-util'
import {
  NownodesProvider,
  UtxoOnlineDataProviders,
} from '@xchainjs/xchain-utxo-providers'

export const MIN_TX_FEE = 10000

export const ZEC_DECIMAL = 8

export const LOWER_FEE_BOUND = 1
export const UPPER_FEE_BOUND = 1_000

// export const ZEC_SYMBOL = '₿'
// export const ZEC_SATOSHI_SYMBOL = '⚡'

/**
 * Chain identifier for Zcash mainnet
 */
export const ZECChain = 'ZEC' as const

/**
 * Base "chain" asset on Zcash main net.
 */
export const AssetZEC: Asset = { chain: ZECChain, symbol: 'ZEC', ticker: 'ZEC', type: AssetType.NATIVE }

// Explorer providers for Zcash
const ZEC_MAINNET_EXPLORER = new ExplorerProvider(
  'https://mainnet.zcashexplorer.app/',
  'https://mainnet.zcashexplorer.app/address/%%ADDRESS%%',
  'https://mainnet.zcashexplorer.app/transactions/%%TX_ID%%',
)
const ZEC_TESTNET_EXPLORER = new ExplorerProvider(
  'https://testnet.zcashexplorer.app/',
  'https://testnet.zcashexplorer.app/address/%%ADDRESS%%',
  'https://testnet.zcashexplorer.app/transactions/%%TX_ID%%',
)
export const blockstreamExplorerProviders = {
  [Network.Testnet]: ZEC_TESTNET_EXPLORER,
  [Network.Stagenet]: ZEC_MAINNET_EXPLORER,
  [Network.Mainnet]: ZEC_MAINNET_EXPLORER,
}

// Bitgo data providers for Zcash
// const mainnetBitgoProvider = new BitgoProvider({
//   baseUrl: 'https://app.bitgo.com',
//   chain: ZECChain,
// })

// export const BitgoProviders: UtxoOnlineDataProviders = {
//   [Network.Testnet]: undefined,
//   [Network.Stagenet]: mainnetBitgoProvider,
//   [Network.Mainnet]: mainnetBitgoProvider,
// }

const mainnetNownodesProvider = new NownodesProvider(
  'https://zecbook.nownodes.io/api/v2',
  ZECChain,
  AssetZEC,
  ZEC_DECIMAL,
  process.env.NOWNODES_API_KEY || ''
)

export const NownodesProviders: UtxoOnlineDataProviders = {
  [Network.Testnet]: undefined,
  [Network.Stagenet]: mainnetNownodesProvider,
  [Network.Mainnet]: mainnetNownodesProvider,
}

