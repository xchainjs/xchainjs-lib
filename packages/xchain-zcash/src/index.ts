// Export the 'ClientKeystore' class as 'Client' for backward compatibility
export { ClientKeystore as Client } from './clientKeystore'

// Also export ClientKeystore by its actual name for explicit usage
export { ClientKeystore } from './clientKeystore'

// Export the base Client class for cases where keystore is not needed
export { Client as BaseClient } from './client'

// Export the 'ClientLedger' class
export { ClientLedger } from './clientLedger'

// Export the 'defaultZECParams' constant from the 'client' file
export { defaultZECParams } from './client'

// Export all constants from the 'const' file
export * from './const'

// Export the 'getPrefix' and 'validateAddress' functions from the 'utils' file
export { getPrefix, validateAddress } from './utils'
