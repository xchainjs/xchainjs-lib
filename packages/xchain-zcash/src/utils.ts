// Import statements for necessary modules and types
import { Network } from '@xchainjs/xchain-client' // Importing the Network type from xchain-client module
import { Address } from '@xchainjs/xchain-util' // Importing the Address type from xchain-util module
import { UTXO } from '@xchainjs/xchain-utxo' // Importing the UTXO type from xchain-utxo module
import * as utxolib from '@bitgo/utxo-lib'

export const TX_EMPTY_SIZE = 4 + 1 + 1 + 4 // Total size of an empty transaction
export const TX_INPUT_BASE = 32 + 4 + 1 + 4 // Size of a base input in a transaction
export const TX_INPUT_PUBKEYHASH = 107 // Size of an input with a public key hash
export const TX_OUTPUT_BASE = 8 + 1 // Size of a base output in a transaction
export const TX_OUTPUT_PUBKEYHASH = 25 // Size of an output with a public key hash

/**
 * Function to calculate the size of an input in a transaction.
 * @param {UTXO} input - The UTXO (Unspent Transaction Output) for which to calculate the size.
 * @returns {number} The size of the input.
 */
export const inputBytes = (input: UTXO): number => { // TODO: Review id this is neccessary for Zcash
  return TX_INPUT_BASE + (input.witnessUtxo?.script ? input.witnessUtxo?.script.length : TX_INPUT_PUBKEYHASH)
}

/**
 * Function to calculate the average value of an array of numbers.
 * @param {number[]} array - The array of numbers.
 * @returns {number} The average value of the array.
 */
export const arrayAverage = (array: number[]): number => {
  let sum = 0
  array.forEach((value) => (sum += value))
  return sum / array.length
}

/**
 * Function to get the Zcash network to be used with @bitgo/utxo-lib.
 *
 * @param {Network} network - The network type (Mainnet, Testnet, or Stagenet).
 * @returns {utxolib.Network} The Zcash network.
 */
export const zecNetwork = (network: Network): utxolib.Network => {
  switch (network) {
    case Network.Mainnet:
    case Network.Stagenet:
      return utxolib.networks.zcash // Return the Zcash mainnet or stagenet network
    case Network.Testnet:
      return utxolib.networks.zcashTest // Return the Zcash testnet network
  }
}

/**
 * Function to validate a Zcash address.
 * @param {Address} address - The Zcash address to validate.
 * @param {Network} network - The network type (Mainnet, Testnet, or Stagenet).
 * @returns {boolean} `true` if the address is valid, `false` otherwise.
 */
export const validateAddress = (address: Address, network: Network): boolean => {
  try {
    utxolib.address.toOutputScript(address, zecNetwork(network)) // Try to convert the address to an output script using the specified network
    return true // If successful, the address is valid
  } catch (error) {
    return false // If an error occurs, the address is invalid
  }
}

/**
 * Function to get the address prefix based on the network.
 * @param {Network} network - The network type (Mainnet, Testnet, or Stagenet).
 * @returns {string} The address prefix based on the network.
 */
export const getPrefix = (network: Network): string => {
  switch (network) {
    case Network.Mainnet:
    case Network.Stagenet:
      return 't1'
    case Network.Testnet:
      return 'tm'
  }
}

/**
 * Converts a public key to an X-only public key.
 * @param pubKey The public key to convert.
 * @returns The X-only public key.
 */
export const toXOnly = (pubKey: Buffer): Buffer => (pubKey.length === 32 ? pubKey : pubKey.subarray(1, 33))
