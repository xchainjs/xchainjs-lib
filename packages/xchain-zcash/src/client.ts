// import * as ecc from '@bitcoin-js/tiny-secp256k1-asmjs'
import { AssetInfo, FeeRate, Network } from '@xchainjs/xchain-client'
import { Address } from '@xchainjs/xchain-util'
import { Client as UTXOClient, PreparedTx, TxParams, UTXO, UtxoClientParams } from '@xchainjs/xchain-utxo'
import * as utxolib from '@bitgo/utxo-lib'
import accumulative from 'coinselect/accumulative'

import {
  AssetZEC,
  ZECChain,
  ZEC_DECIMAL,
  NownodesProviders,
  LOWER_FEE_BOUND,
  MIN_TX_FEE,
  UPPER_FEE_BOUND,
  blockstreamExplorerProviders,
} from './const'
import * as Utils from './utils'

// Default parameters for the Zcash client
export const defaultZECParams: UtxoClientParams = {
  network: Network.Mainnet,
  phrase: '',
  explorerProviders: blockstreamExplorerProviders,
  dataProviders: [NownodesProviders],
  rootDerivationPaths: {
    [Network.Mainnet]: `m/44'/133'/0'/0/`,
    [Network.Testnet]: `m/44'/1'/0'/0/`,
    [Network.Stagenet]: `m/44'/133'/0'/0/`,
  },
  feeBounds: {
    lower: LOWER_FEE_BOUND,
    upper: UPPER_FEE_BOUND,
  },
}
/**
 * Custom Zcash client (only support t-addresses)
 */
abstract class Client extends UTXOClient {
  /**
   * Constructor
   * Initializes the client with network type and other parameters.
   * @param {UtxoClientParams} params
   */
  constructor(
    params: UtxoClientParams = {
      ...defaultZECParams,
    },
  ) {
    super(ZECChain, {
      network: params.network,
      rootDerivationPaths: params.rootDerivationPaths,
      phrase: params.phrase,
      feeBounds: params.feeBounds,
      explorerProviders: params.explorerProviders,
      dataProviders: params.dataProviders,
    })
  }

  /**
   * Get ZEC asset info.
   * @returns {AssetInfo} ZEC asset information.
   */
  getAssetInfo(): AssetInfo {
    const assetInfo: AssetInfo = {
      asset: AssetZEC,
      decimal: ZEC_DECIMAL,
    }
    return assetInfo
  }

  /**
   * Validate the given Zcash address.
   * @param {string} address Zcash address to validate (only t-addresses).
   * @returns {boolean} `true` if the address is valid, `false` otherwise.
   */
  validateAddress(address: string): boolean {
    return Utils.validateAddress(address, this.network)
  }

  /**
   * Compile memo into a buffer.
   * @param {string} memo Memo to compile.
   * @returns {Buffer} Compiled memo.
   */
  protected compileMemo(memo: string): Buffer {
    const data = Buffer.from(memo, 'utf8') // converts MEMO to buffer
    return utxolib.script.compile([utxolib.opcodes.OP_RETURN, data]) // Compile OP_RETURN script
  }

  /**
   * Get transaction fee from UTXOs.
   * @param {UTXO[]} inputs UTXOs to calculate fee from.
   * @param {FeeRate} feeRate Fee rate.
   * @param {Buffer | null} data Compiled memo (Optional).
   * @returns {number} Transaction fee.
   */
  protected getFeeFromUtxos(inputs: UTXO[], feeRate: FeeRate, data: Buffer | null = null): number {
    // Calculate input size based on inputs
    const inputSizeBasedOnInputs =
      inputs.length > 0
        ? inputs.reduce((a, x) => a + Utils.inputBytes(x), 0) + inputs.length // +1 byte for each input signature
        : 0
    // Calculate sum
    let sum =
      Utils.TX_EMPTY_SIZE +
      inputSizeBasedOnInputs +
      Utils.TX_OUTPUT_BASE +
      Utils.TX_OUTPUT_PUBKEYHASH +
      Utils.TX_OUTPUT_BASE +
      Utils.TX_OUTPUT_PUBKEYHASH

    if (data) {
      sum += Utils.TX_OUTPUT_BASE + data.length
    }
    // Calculate fee
    const fee = sum * feeRate
    return fee > MIN_TX_FEE ? fee : MIN_TX_FEE
  }

  async getUtxo(address: string) {
    return await this.scanUTXOs(address, true)
  }

  /**
   * Build a Zcash transaction.*
   * @param param0
   * @deprecated
   */
  async buildTx({
    amount,
    recipient,
    memo,
    feeRate,
    sender,
    spendPendingUTXO = true,
  }: TxParams & {
    feeRate: FeeRate
    sender: Address
    spendPendingUTXO?: boolean
    withTxHex?: boolean
  }): Promise<{ txb: utxolib.bitgo.ZcashTransactionBuilder; utxos: UTXO[]; inputs: UTXO[] }> {
    // Check memo length
    if (memo && memo.length > 80) {
      throw new Error('memo too long, must not be longer than 80 chars.')
    }
    // This section of the code is responsible for preparing a transaction by building a Bitcoin PSBT (Partially Signed Bitcoin Transaction).
    if (!this.validateAddress(recipient)) throw new Error('Invalid address')
    // Determine whether to only use confirmed UTXOs or include pending UTXOs based on the spendPendingUTXO flag.
    const confirmedOnly = !spendPendingUTXO
    // Scan UTXOs associated with the sender's address.
    const utxos = await this.scanUTXOs(sender, confirmedOnly)
    // Throw an error if there are no available UTXOs to cover the transaction.
    if (utxos.length === 0) throw new Error('Insufficient Balance for transaction')
    // Round up the fee rate to the nearest integer.
    const feeRateWhole = Math.ceil(feeRate)
    // Compile the memo into a Buffer if provided.
    const compiledMemo = memo ? this.compileMemo(memo) : null
    // Initialize an array to store the target outputs of the transaction.
    const targetOutputs = []

    // 1. Add the recipient address and amount to the target outputs.
    targetOutputs.push({
      address: recipient,
      value: amount.amount().toNumber(),
    })
    // 2. Add the compiled memo to the target outputs if it exists.
    if (compiledMemo) {
      targetOutputs.push({ script: compiledMemo, value: 0 })
    }
    // Use the coinselect library to determine the inputs and outputs for the transaction.
    const { inputs, outputs } = accumulative(utxos, targetOutputs, feeRateWhole)
    // If no suitable inputs or outputs are found, throw an error indicating insufficient balance.
    if (!inputs || !outputs) throw new Error('Insufficient Balance for transaction')
    // Initialize a new Bitcoin PSBT object.
    let txb = new utxolib.bitgo.ZcashTransactionBuilder(Utils.zecNetwork(this.network))

    txb.setVersion(5)
    txb.setVersionGroupId(0x26a7270a)
    txb.setExpiryHeight(0)
    txb.setConsensusBranchId(0xc8e71055)

    inputs.forEach((utxo: UTXO) => {
      // if (!utxo.txHex) {
      //   throw Error (`Non txHex data for UTXO ${utxo.txHex}`)
      // }
      // const tx = utxolib.Transaction.fromHex(utxo.txHex)
      // console.log('tx', tx)
      // txHash: Buffer | string | Transaction<TNumber>, vout: number, sequence?: number, prevOutScript?: Buffer, value?: TNumber
      txb.addInput(
        utxo.hash,
        utxo.index,
        undefined,
        Buffer.from(utxo.scriptPubKey as string, 'hex'),
        utxo.value
      )
    }
    )
    // Add outputs to the PSBT from the accumulated outputs.
    outputs.forEach((output: any) => {
      // If the output address is not specified, it's considered a change address and set to the sender's address.
      if (!output.address) {
        //an empty address means this is the change address
        output.address = sender
      }
      // Add the output to the PSBT.
      if (!output.script) {
        txb.addOutput(output.address, output.value)
      } else {
        // If the output is a memo, add it to the PSBT to avoid dust error.
        if (compiledMemo) {
          txb.addOutput(compiledMemo, 0)
        }
      }
    })

    return { txb: txb, utxos, inputs }
  }

  /**
   * Prepare transfer.
   *
   * @param {TxParams&Address&FeeRate&boolean} params The transfer options.
   * @returns {PreparedTx} The raw unsigned transaction.
   */
  async prepareTx({
    sender,
    memo,
    amount,
    recipient,
    spendPendingUTXO = true,
    feeRate,
  }: TxParams & {
    sender: Address
    feeRate: FeeRate
    spendPendingUTXO?: boolean
  }): Promise<PreparedTx> {
    // Build the transaction using the provided parameters.
    const { txb, utxos, inputs } = await this.buildTx({
      sender,
      recipient,
      amount,
      feeRate,
      memo,
      spendPendingUTXO,
    })
    // Return the raw unsigned transaction (PSBT) and associated UTXOs.
    return { rawUnsignedTx: txb.buildIncomplete().toHex(), utxos, inputs }
  }
}

export { Client }
