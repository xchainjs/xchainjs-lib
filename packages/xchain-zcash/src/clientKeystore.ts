import { FeeOption, FeeRate, TxHash, checkFeeBounds } from '@xchainjs/xchain-client'
import { getSeed } from '@xchainjs/xchain-crypto'
import { Address } from '@xchainjs/xchain-util'
import { TxParams, UtxoClientParams } from '@xchainjs/xchain-utxo'
import * as utxolib from '@bitgo/utxo-lib'
import { createHash } from 'crypto'
import * as bs58check from 'bs58check'
import { BIP32Factory } from 'bip32'
import * as ecc from '@bitcoin-js/tiny-secp256k1-asmjs'

import { Client, defaultZECParams } from './client' // Importing the base Bitcoin client
import * as Utils from './utils'

/**
 * Custom Bitcoin client extended to support keystore functionality
 */
class ClientKeystore extends Client {
  constructor(
    params: UtxoClientParams = {
      ...defaultZECParams,
    },
  ) {
    super(params)
  }
  /**
   * @deprecated This function eventually will be removed. Use getAddressAsync instead.
   * Get the address associated with the given index.
   * @param {number} index The index of the address.
   * @returns {Address} The Bitcoin address.
   * @throws {"index must be greater than zero"} Thrown if the index is less than zero.
   * @throws {"Phrase must be provided"} Thrown if the phrase has not been set before.
   * @throws {"Address not defined"} Thrown if failed to create the address from the phrase.
   */
  getAddress(index = 0): Address {
    // Check if the index is valid
    if (index < 0) {
      throw new Error('index must be greater than zero')
    }

    // Check if the phrase has been set
    if (this.phrase) {
      // const zecNetwork = Utils.zecNetwork(this.network)
      const zecKeys = this.getZecKeys(this.phrase, index)

      // let address: string | undefined
      // address = utxolib.payments.p2wpkh({
      //   pubkey: zecKeys.publicKey,
      //   network: zecNetwork,
      // }).address
      // // Throw an error if the address is not defined
      // if (!address) {
      //   throw new Error('Address not defined')
      // }

      // return address

      // const keyPair = utxolib.ECPair.fromPrivateKey(node.privateKey)

      const sha256 = createHash('sha256').update(zecKeys.publicKey).digest()
      const ripemd160 = createHash('ripemd160').update(sha256).digest()
      const prefix = Buffer.from([0x1c, 0xb8]) // ZEC P2PKH
      const payload = Buffer.concat([prefix, ripemd160])

      return bs58check.default.encode(payload)
    }

    throw new Error('Phrase must be provided')
  }

  /**
   * Get the current address asynchronously.
   * @param {number} index The index of the address.
   * @returns {Promise<Address>} A promise that resolves to the Bitcoin address.
   * @throws {"Phrase must be provided"} Thrown if the phrase has not been set before.
   */
  async getAddressAsync(index = 0): Promise<string> {
    return this.getAddress(index)
  }

  /**
   * @private
   * Get the keys derived from the given phrase.
   *
   * @param {string} phrase The phrase to be used for generating the keys.
   * @param {number} index The index of the address.
   * @returns {Bitcoin.ECPair.ECPairInterface} The Bitcoin key pair.
   * @throws {"Could not get private key from phrase"} Thrown if failed to create BTC keys from the given phrase.
   */
  private getZecKeys(phrase: string, index = 0): utxolib.ECPairInterface {
    // const zecNetwork = Utils.zecNetwork(this.network)
    const seed = getSeed(phrase)

    const bip32 = BIP32Factory(ecc)
    const master = bip32.fromSeed(seed).derivePath(this.getFullDerivationPath(index))

    if (!master.privateKey) {
      throw new Error('Could not get private key from phrase')
    }

    return utxolib.ECPair.fromPrivateKey(Buffer.from(master.privateKey)) // Be carefull missing zcash network due to this error: https://github.com/iancoleman/bip39/issues/94 
  }

  /**
   * Transfer ZEC.
   *
   * @param {TxParams&FeeRate} params The transfer options including the fee rate.
   * @returns {Promise<TxHash|string>} A promise that resolves to the transaction hash or an error message.
   * @throws {"memo too long"} Thrown if the memo is longer than 80 characters.
   */
  async transfer(params: TxParams & { feeRate?: FeeRate }): Promise<TxHash> {
    // Set the default fee rate to `fast`
    const feeRate = params.feeRate || (await this.getFeeRates())[FeeOption.Fast]

    // Check if the fee rate is within the fee bounds
    checkFeeBounds(this.feeBounds, feeRate)

    // Get the address index from the parameters or use the default value
    const fromAddressIndex = params?.walletIndex || 0

    // Get the Bitcoin keys
    const zecKeys = this.getZecKeys(this.phrase, fromAddressIndex)
    const sender = await this.getAddressAsync(fromAddressIndex)
    // zecKeys.network = Utils.zecNetwork(this.network)

    // Prepare the transaction
    const { rawUnsignedTx, inputs } = await this.prepareTx({
      ...params,
      sender,
      feeRate,
    })

    // Build the PSBT
    const buffer = Buffer.from(rawUnsignedTx, 'hex')
    const unsignedTx = utxolib.bitgo.ZcashTransaction.fromBuffer(buffer, false, 'number', Utils.zecNetwork(this.network))
    const txb = utxolib.bitgo.ZcashTransactionBuilder.fromTransaction(unsignedTx, Utils.zecNetwork(this.network))

    // Sign all inputs
    for (let i = 0; i < inputs.length; i++) {
      // txb.tx.ins[i] = { ...txb.tx.ins[i], value: inputs[i].value } as any
      txb.sign(
        inputs[i].index,
        zecKeys,
        undefined,
        utxolib.Transaction.SIGHASH_ALL,
        inputs[i].value,
        // value: inputs[i].value
      )
    }

    // Extract the transaction hex
    const txBuilded = txb.build()
    const txHex = txBuilded.toHex()
    // Extract the transaction hash
    // const txHash = psbt.extractTransaction().getId()

    const txId = await this.roundRobinBroadcastTx(txHex)
    return txId
  }
}

export { ClientKeystore }
