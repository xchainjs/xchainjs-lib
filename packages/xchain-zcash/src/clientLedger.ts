import { TxHash } from '@xchainjs/xchain-client'
import { Address } from '@xchainjs/xchain-util'
import { UtxoClientParams } from '@xchainjs/xchain-utxo'

import { Client } from './client'

class ClientLedger extends Client {

  constructor(params: UtxoClientParams) {
    super(params)
    throw Error('Ledger client not supported for Zcash.')
  }

  public async getApp(): Promise<any> {
    throw Error('Not implemented.')
  }

  getAddress(): string {
    throw Error('Not implemented.')
  }

  async getAddressAsync(): Promise<Address> {
    throw Error('Not implemented.')
  }

  async transfer(): Promise<TxHash> {
    throw Error('Not implemented.')
  }
}

export { ClientLedger }
