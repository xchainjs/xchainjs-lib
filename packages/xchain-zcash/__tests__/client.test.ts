import { Client, defaultZECParams } from '../src'
// import MockNowNodes from '../__mocks__/nownodes'
import * as bip39 from 'bip39'

function deepSerialize(obj: any): any {
  if (Array.isArray(obj)) {
    return obj.map(deepSerialize)
  }

  if (obj instanceof Date) {
    return obj.toISOString()
  }

  if (obj && typeof obj === 'object') {
    if (typeof obj.amount === 'function' && 'decimal' in obj && 'type' in obj) {
      return {
        value: obj.amount().toString(),
        decimal: obj.decimal,
        type: obj.type,
      }
    }

    const serialized: any = {}
    for (const key in obj) {
      serialized[key] = deepSerialize(obj[key])
    }
    return serialized
  }

  // Casos base: strings, numbers, booleans, null...
  return obj
}

describe('Zcash client', () => {
  let client: Client

  beforeAll(() => {
    const mnemonic = bip39.entropyToMnemonic('00000000000000000000000000000000');
    client = new Client({
      ...defaultZECParams,
      phrase: mnemonic,
    })
  })

  // beforeEach(() => {
  //   MockNowNodes.init()
  // })

  // afterEach(() => {
  //   MockNowNodes.restore()
  // })

  it('Should generate address', async () => {
    const address = await client.getAddressAsync()
    expect(address).toBe('t1XVXWCvpMgBvUaed4XDqWtgQgJSu1Ghz7F')
  })

  it('Should get balance', async () => {
    const balance = await client.getBalance('t1eiZYPXWurGMxFwoTu62531s8fAiExFh88')
    expect(balance[0].amount.amount().toString()).toBe("268682490")
    expect(balance[0].asset).toStrictEqual({"chain": "ZEC", "symbol": "ZEC", "ticker": "ZEC", "type": 0})
  })

  it('Should get tx', async () => {
    const tx = await client.getTransactionData('601ead82c513ec0b0e125f86ec5ff50d115a0525786e5d10de38a9c95f61b789')
    const txDeserialize = deepSerialize(tx)
    expect(txDeserialize).toMatchSnapshot('zcash-get-tx')
  })

  it('Should get txs', async () => {
    const txs = await client.getTransactions({
      address: 't1eiZYPXWurGMxFwoTu62531s8fAiExFh88'
    })
    const txsDeserialize = deepSerialize(txs)
    expect(txsDeserialize).toMatchSnapshot('zcash-get-txs')
  })

  it('Should get UTXO confirmed', async () => {
    const utxo = await client.getUtxo('t1eiZYPXWurGMxFwoTu62531s8fAiExFh88')
    const txsDeserialize = deepSerialize(utxo)
    expect(txsDeserialize).toMatchSnapshot('zcash-get-utxo-confirmed')
  })

})
