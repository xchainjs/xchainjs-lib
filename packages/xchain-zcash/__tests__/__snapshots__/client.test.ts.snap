// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zcash client Should get UTXO confirmed: zcash-get-utxo-confirmed 1`] = `
[
  {
    "hash": "601ead82c513ec0b0e125f86ec5ff50d115a0525786e5d10de38a9c95f61b789",
    "index": 0,
    "value": 268682490,
  },
]
`;

exports[`Zcash client Should get tx: zcash-get-tx 1`] = `
{
  "asset": {
    "chain": "ZEC",
    "symbol": "ZEC",
    "ticker": "ZEC",
    "type": 0,
  },
  "date": "2025-04-08T17:20:05.000Z",
  "from": [
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "95766509",
      },
      "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "39081200",
      },
      "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "35494039",
      },
      "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "15804625",
      },
      "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "12839186",
      },
      "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "1031041542",
      },
      "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
  ],
  "hash": "601ead82c513ec0b0e125f86ec5ff50d115a0525786e5d10de38a9c95f61b789",
  "to": [
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "268682490",
      },
      "to": "t1eiZYPXWurGMxFwoTu62531s8fAiExFh88",
    },
    {
      "amount": {
        "decimal": 8,
        "type": "BASE",
        "value": "961303931",
      },
      "to": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
    },
  ],
  "type": "transfer",
}
`;

exports[`Zcash client Should get txs: zcash-get-txs 1`] = `
{
  "total": 1,
  "txs": [
    {
      "asset": {
        "chain": "ZEC",
        "symbol": "ZEC",
        "ticker": "ZEC",
        "type": 0,
      },
      "date": "2025-04-08T17:20:05.000Z",
      "from": [
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "95766509",
          },
          "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "39081200",
          },
          "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "35494039",
          },
          "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "15804625",
          },
          "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "12839186",
          },
          "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "1031041542",
          },
          "from": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
      ],
      "hash": "601ead82c513ec0b0e125f86ec5ff50d115a0525786e5d10de38a9c95f61b789",
      "to": [
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "268682490",
          },
          "to": "t1eiZYPXWurGMxFwoTu62531s8fAiExFh88",
        },
        {
          "amount": {
            "decimal": 8,
            "type": "BASE",
            "value": "961303931",
          },
          "to": "t1Qfn52D2TGCAC75aBEKiWkozeJgaQsR638",
        },
      ],
      "type": "transfer",
    },
  ],
}
`;
